ENH: add setters/deleters for Sequence.metadata/positional_metadata

jairideout opened the pull request:
Other changes:
  - Added shallow copying of metadata/positional metadata and docs describing this behavior.
  - Positional metadata index is now always set to ``np.arange(len(self))``.
  - Added ``skbio.util.assert_data_frame_almost_equal`` and updated unit tests to use it.

Address part of #949. Fixes #948.

jairideout commented:
@ebolyen @anderspitman can you both review this? I think this one's worth having a couple set of eyes on.

ebolyen commented on skbio/sequence/_sequence.py:
```diff
@@ -106,20 +99,181 @@ class Sequence(collections.Sequence, SkbioObject):
+    ``pd.DataFrame``:
+
+    >>> positional_metadata = pd.DataFrame(
+    ...     {'qual': [3, 3, 4, 10], 'list': [[], [], [], []]})
```
Change this to `quality` since that is what `io` will be producing

jairideout replied:
done

ebolyen commented on skbio/sequence/_sequence.py:
```diff
@@ -106,20 +99,181 @@ class Sequence(collections.Sequence, SkbioObject):
+    >>> subseq.positional_metadata
+      list  qual   gaps
+    0   []     3  False
+    1   []     4  False
```
I think there should be an example of setting a non-mutable piece of metadata to make it clear that only mutable objects have this behavior, simple things like ints and strings are fine.

jairideout replied:
I have examples of this above where I add a new columns (`gaps`) and change one of the quality scores to 999.

ebolyen replied:
Right, sorry, I apparently can't ready RST without my eyes glazing over.

jairideout replied:
I know the feeling

ebolyen commented on skbio/sequence/_sequence.py:
```diff
@@ -106,20 +99,181 @@ class Sequence(collections.Sequence, SkbioObject):
     """
     default_write_format = 'fasta'
-    __hash__ = None  # TODO revisit hashability when all properties are present
+    __hash__ = None
```
:+1:

ebolyen commented on skbio/sequence/_sequence.py:
```diff
@@ -187,30 +351,107 @@ def metadata(self):
+
+    @metadata.deleter
+    def metadata(self):
+        if hasattr(self, '_metadata'):
```
Why is this check necessary? Can't you just reassign instead?

jairideout replied:
`del` gets called from the constructor where `self._metadata` hasn't been assigned yet. I did it this way to have only a single representation of "missing" metadata (defined by the deleter). I have a comment about this in the constructor but I'll also add one here for clarity.

jairideout replied:
done

jairideout replied:
With efficient handling of missing metadata this code was refactored. It's much clearer now, both here and in the constructor.

ebolyen commented on skbio/sequence/_sequence.py:
```diff
@@ -187,30 +351,107 @@ def metadata(self):
+
+        Update positional metadata:
+
+        >>> s.positional_metadata['gaps'] = [False, False, False, False]
```
Would it be worth using `DNA` for this example so that you could use `s.positional_metadata['gaps'] = s.gaps()`?

jairideout replied:
good idea, updated to use `.gaps()` and `.degenerates()`

ebolyen commented on skbio/sequence/_sequence.py:
```diff
@@ -187,30 +351,107 @@ def metadata(self):
+    @positional_metadata.deleter
+    def positional_metadata(self):
+        if hasattr(self, '_positional_metadata'):
+            del self._positional_metadata
```
Same question as above, is this needed?

jairideout replied:
done

jairideout replied:
same comment as above

ebolyen commented on skbio/sequence/_sequence.py:
```diff
@@ -187,30 +351,107 @@ def metadata(self):
+    def positional_metadata(self):
+        if hasattr(self, '_positional_metadata'):
+            del self._positional_metadata
+        self.positional_metadata = pd.DataFrame(index=np.arange(len(self)))
```
Any chance this can be optimized to be `self._positional_metadata = None` and the getter can just build a dataframe on the fly? You shouldn't have to change any tests

jairideout replied:
you got it

ebolyen commented on skbio/sequence/_sequence.py:
```diff
@@ -223,13 +464,19 @@ def __init__(self, sequence, metadata=None,
+
+        if metadata is None:
+            # reset to default empty metadata representation
+            del self.metadata
```
I know this is being optimistic, but you could also do `self._metadata = None` and modify the getter slightly

jairideout replied:
done

ebolyen commented on skbio/sequence/_sequence.py:
```diff
@@ -223,13 +464,19 @@ def __init__(self, sequence, metadata=None,
+            self.metadata = metadata
+
+        if positional_metadata is None:
+            del self.positional_metadata
```
Same here

jairideout replied:
done

ebolyen commented:
Looks good @jairideout, just a couple minor comments in the deleters and a request for a memory optimization when metadata isn't needed (should be simple and won't impact tests).

anderspitman commented on skbio/sequence/_sequence.py:
```diff
@@ -187,30 +351,107 @@ def metadata(self):
     @property
     def positional_metadata(self):
-        """Pandas DataFrame containing metadata on a per-character basis.
+        """pandas.DataFrame containing metadata on a per-character basis.
```
[Here](https://github.com/biocore/scikit-bio/pull/954/files#diff-4ac32a78649ca5bdd8e0ba38b7006a1eR13) you use pd.DataFrame instead of pandas.DataFrame. Should probably be consistent.

ebolyen replied:
I like `pd.DataFrame`

jairideout replied:
:+1: I'll go with `pd.DataFrame` since we're generally using `np.something` in the sequence docs

jairideout replied:
done

anderspitman replied:
:+1:

anderspitman commented on skbio/sequence/tests/test_sequence.py:
```diff
@@ -82,8 +88,15 @@ def test_init_empty_sequence(self):
             npt.assert_equal(seq.values, np.array('', dtype='c'))
-            self.assertEqual('', str(seq))
-            self.assertEqual(0, len(seq))
+            self.assertEqual(str(seq), '')
```
Is this order (observed, expected) our convention in general or did you just change it to be consistent with the neighboring code?

ebolyen replied:
generally yes, on both accounts.

jairideout replied:
yes

anderspitman commented:
Done with review. Looks really good overall.

jairideout commented:
@ebolyen I added the memory optimization you requested for `metadata` and `positional_metadata` (#948). I updated all internal usage of these properties to also be memory efficient. Finally we have tests that should catch regression in the future. Affected methods were `__eq__`, `__getitem__`, `has_metadata`, `has_positional_metadata`, `__init__`, and `_to`. The getter/setter/deleter code is actually cleaner and easier to follow now!

@ebolyen @anderspitman this is ready for another review (there were substantial changes in my latest commit).

ebolyen commented on skbio/io/tests/test_fasta.py:
```diff
@@ -411,8 +411,10 @@ def test_fasta_to_generator_valid_files(self):
-                                             ignore=['positional_metadata']))
+                    # ignore positional metadata by creating a copy of the
+                    # expected sequence without positional metadata
+                    e = e._to(positional_metadata=None)
```
Can you add a TODO saying to use `copy` and a setter in the future?

ebolyen replied:
Actually I see this is used a lot, maybe it's not worth it to worry about it now.

jairideout replied:
Added a TODO to this occurrence. When we have `copy` we can search for all usages of `_to` and replace with `copy` and a setter.

ebolyen commented on skbio/sequence/_sequence.py:
```diff
@@ -180,37 +334,128 @@ def metadata(self):
+
+    @positional_metadata.deleter
+    def positional_metadata(self):
+        del self._positional_metadata
```
Is this still needed? The GC should be all we need now.

jairideout replied:
fixed

ebolyen commented on skbio/sequence/_sequence.py:
```diff
@@ -180,37 +334,128 @@ def metadata(self):
+
+    @metadata.deleter
+    def metadata(self):
+        del self._metadata
```
Is this still needed?

jairideout replied:
fixed

ebolyen commented:
@jairideout finished with review, just two questions, then I think this is good to go!

jairideout commented:
Thanks for reviewing @ebolyen! I addressed your comments.

ebolyen commented:
Thanks @jairideout!
